### DISCUSSION TITLE
Fields feedback

### DISCUSSION THREAD

**Discussion by adrianthedev** [avo-hq/avo#834] [Feature feedback]
Let's use this space to provide feedback for anything field-related.

Post [an issue](https://github.com/avo-hq/avo/issues/new/choose) if something is broken.

**Comment by Sandilly** [comment]
Hi, I am trying to create a field that will automatically generate a default value off another form field input value. Does the default block have access to other form values on create (values that have not been saved to the record)?

**Reply by adrianthedev** [reply]
You can do that using the StimulusJS integration.
You'll need to write a controller that watches for changes and updates the field.

Docs with demo video:
https://docs.avohq.io/3.0/stimulus-integration.html


Live demo:
https://avodemo.com/avo/resources/courses/147/edit

The code:
https://github.com/avo-hq/main.avodemo.com/blob/main/app/javascript/controllers/course_controller.js

**Reply by Sandilly** [reply]
Thank you for your guidance. I will try this.

**Comment by manuelmeurer** [comment]
I added a field and it doesn't show up in the index or show view:

```ruby
class Avo::Resources::Talk < Avo::BaseResource
  def fields
    field :speakers, as: :has_many
  end
end
```

This is my model:

```ruby
class Talk < ApplicationRecord
  has_many :talk_speakers
  has_many :speakers, through: :talk_speakers, source: :user
end
```

Question 1: what do I have to do to make it show up in the show view (not the index view)?
Question 2: if i misconfigured something, why does the field simply not show up instead of raising an error? It's super frustrating to not get any feedback from the system about what might be wrong.

**Reply by Paul-Bob** [reply]
Hi @manuelmeurer, thanks for reaching out.

> I added a field and it doesn't show up in the index or show view

I see that you added a `has_many` field, but it isn't appearing anywhere. By default, the `has_many` field is only visible on the Show view. If you want it to appear on the Edit view as well, you can add `show_on: :edit` to the field's configuration. You can find more details on this [documentation page](https://docs.avohq.io/3.0/associations/has_many.html).

Your resource and model seem to be correctly configured. I don’t believe this is the cause of your issue.
, however, since the field is a `has_many` `through`, I recommend explicitly configuring the `through` option in the resource ([documented here](https://docs.avohq.io/3.0/associations/has_many.html#has-many-through)).

--- 

> Question 1: what do I have to do to make it show up in the show view (not the index view)?

`has_many` fields should be visible in the Show view by default. However, if you're using authorization, you need to explicitly allow the field to be displayed. In your case, you should add `def show_speakers? = true` to the Talk policy file. More details can be found in the [authorization documentation](https://docs.avohq.io/3.0/authorization.html#view_association).

---

> Question 2: if i misconfigured something, why does the field simply not show up instead of raising an error? It's super frustrating to not get any feedback from the system about what might be wrong.

I understand the frustration. We're continuously working to improve the development experience by adding custom alerts and raising meaningful errors in such cases. If authorization is the issue, there should be logs indicating the unauthorized action. You can refer to the [authorization logs documentation](https://docs.avohq.io/3.0/authorization.html#logs) for more details.

If it’s something else, no worries, we’ll track it down and, if it makes sense, add a system notification for it

---

To summarize, `has_many` fields should be visible by default in the Show view. If you're using authorization, you need to explicitly allow the field to be displayed.

Let us know if authorization is preventing the field from rendering or if there's another issue that requires further debugging,

Thank you!

**Reply by manuelmeurer** [reply]
Adding `view_speakers?` to the talk policy solved this:

```
class TalkPolicy < ApplicationPolicy
  def view_speakers? = true

  class Scope < ApplicationPolicy::Scope
  end
end
```

Probably an error should be raised if the resource has a `:has_many` field and the policy does not have a `view_resources?` method?

**Reply by manuelmeurer** [reply]
Another question: my Talk model has a JSON field named "tracks" that contains an array of strings. I'd like to show this field, either as a list with bullets or comma joined, for each talk in the show and index views. How can I do this?
I added `field :tracks, as: :array` to the resource, but again, the field doesn't show up anywhere. No warning or error in the logs either.

**Reply by Paul-Bob** [reply]
> Probably an error should be raised if the resource has a `:has_many` field and the policy does not have a `view_resources?` method?

We’re considering adding notifications to give developers an overview of what is being unauthorized when the page loads.

---

> I added field :tracks, as: :array to the resource, but again, the field doesn't show up anywhere. No warning or error in the logs either.

Got it! Array fields and array resources are currently in beta. Simply adding `field :tracks, as: :array` won’t work out of the box. You need to define a dedicated resource for `tracks` and specify which fields should be displayed within that resource. You can find more details in the [documentation](https://docs.avohq.io/3.0/fields/array.html).

Even if you have already created the resource, the field still might not appear. The underlying issue is that array fields inherit most behaviors from `has_many` fields. As a result, you need to explicitly allow viewing by adding `def view_tracks? = true` to your policy.

This requirement isn’t currently documented, but I’ll make sure we update the docs to mention it. Thanks for bringing this up, it helps improve the beta phase of array fields!


---

> Another question: my Talk model has a JSON field named "tracks" that contains an array of strings. I'd like to show this field, either as a list with bullets or comma joined, for each talk in the show and index views. How can I do this?

Since you want to display an array of values in the index and show views, I recommend using the [tags](https://docs.avohq.io/3.0/fields/tags.html) field type. This will automatically process the array and display each item as a tag.

```ruby
# app/models/talk.rb
class Talk < ApplicationRecord
  def tracks
    ["one", "two", "three"]
  end
end

# app/avo/resources/talk.rb
class Avo::Resources::Talk < Avo::BaseResource
  def fields
    # only on the show and index pages
    field :tracks, as: :tags, only_on: :display
  end
end
```

This should render the `tracks` values as styled tags, making them more visually appealing and readable.

![image](https://github.com/user-attachments/assets/81cc68f9-e4db-43f5-b3cf-f86d0c28e925)

This is just one approach, feel free to explore other solutions and let us know how it works for you! You can also create a [custom field](https://docs.avohq.io/3.0/custom-fields.html) if you need greater flexibility over the design and behavior of the `tracks` section. Custom fields let you fine-tune the presentation to better fit your specific requirements.

**Reply by manuelmeurer** [reply]
The "tags" field type is exactly what I was looking for, thanks! :)

**Reply by Paul-Bob** [reply]
Awesome! I'm glad it worked out for your use-case.

Anytime! Thanks for all the feedback!

**Comment by manuelmeurer** [comment]
Here's another issue where I'm a bit stumped that it doesn't work, and there's no error message and no log output.
I'm trying to add a [dynamic filter](https://docs.avohq.io/3.0/dynamic-filters.html) to a field.

```ruby
class Avo::Resources::Ticket < Avo::BaseResource
  def fields
    field :item_name, as: :select, options: -> { record.item_names }, filterable: true
  end
end
```

```ruby
class Ticket < ApplicationRecord
  def self.ransackable_attributes(auth_object = nil) = %w(item_name)
  def item_names
    raise
  end
end
```

The "Filters" button does not appear and nothing is raised, i.e. the `Ticket#item_names` method is never called.
What am I doing wrong?

**Reply by manuelmeurer** [reply]
I'm on Avo Pro 3.17.9, in case that makes a difference.

**Reply by Paul-Bob** [reply]
Hi @manuelmeurer 

> I'm on Avo Pro 3.17.9, in case that makes a difference.

Dynamic filters are included on the Advanced plan that's why `filterable` option takes no effect on Pro or Community plan.

Thank you for the feedback!

**Reply by manuelmeurer** [reply]
Oh, that's disappointing. I think "dynamic filters" should be the default, I needed those in every admin CMS I ever worked with.
Also, if a user on the community or pro plan uses the parameter, I would expect a warning about it in the UI or logs.

**Reply by adrianthedev** [reply]
Hey Manuel, this is Adrian, the author of Avo.

I understand. I really do. I use filters in most of my apps as well. That's why we have the [basic filters](https://docs.avohq.io/3.0/basic-filters.html) in the Community tier for free.

The Dynamic Filters are incredibly powerful and enable developers to create a richer experience for their users. Also it requires a higher level of support form us to capture all the corner cases. That's the reason why they are in the Advanced tier.

@Paul-Bob has created [an issue](https://github.com/avo-hq/avo/issues/3745) to add more warnings for paid features.

THanks for the feedback ✌️

**Reply by manuelmeurer** [reply]
Thanks for your reply, @adrianthedev! I respectfully disagree, since I don't think a feature that pretty much every CMS needs at some point should be in the $150/month price range, but hey, in the end it's your decision. 😄 
I still get a lot of value out of the Pro version of Avo.

**Comment by manuelmeurer** [comment]
Another question about field visibility: I added a `text` field and it showed up in the index and show views. I changed it to `textarea` and now it is only displayed on show views. Fair enough, I understand why that should be the default.
In my case, I'd like to display a (truncated) version of the field on the index view as well.
How can I do that? I tried `show_on: :index`, `show_on: :display` and `show_on: :all`, all to no effect.
I feel like this should be explained on https://docs.avohq.io/3.0/fields/textarea.html (both that textareas are hidden on index views by default, and how to get them to be displayed).

**Reply by Paul-Bob** [reply]
Hi @manuelmeurer, thanks for the feedback!

You're correct, `textarea` fields in Avo don't have an `index` component, so forcing `show_on` won't affect their visibility on the index view.

The common pattern to handle this is to define two fields:
- One for the full `textarea` (for show/edit views),
- Another `text` field specifically for the index, where you can show a sanitized and truncated excerpt.

Here's an example:

```ruby
# app/avo/resources/review.rb
class Avo::Resources::Review < Avo::BaseResource
  def fields
    field :body, as: :textarea
    field :excerpt, as: :text, only_on: :index do
      ActionView::Base.full_sanitizer.sanitize(record.body.to_s).truncate 60
    rescue
      ""
    end
  end
end
```

It be worth mentioning this in the [textarea field documentation](https://docs.avohq.io/3.0/fields/textarea.html) for clarity, as you suggested.

**Reply by manuelmeurer** [reply]
Nice thanks, I'll do it like that!

Maybe this should be the default though? :)
I.e. textareas should have an index component that shows a truncated and sanitized excerpt?

**Reply by Paul-Bob** [reply]
I thought about that while writing the answer, and it makes sense. Right now, introducing it as a default would cause a "breaking change," since all apps upgrading would suddenly have an excerpt displayed on the index and that might not always be desired.

It's definitely something to consider for Avo 4.

In the meantime, you can create a custom index component for the `textarea` field yourself and set it as the default behavior in your application.

Let me know if you need any assistance in that direction.

**Reply by manuelmeurer** [reply]
Cool, can you point me to the docs about a custom index component?

**Reply by manuelmeurer** [reply]
Hi @Paul-Bob, where can I find info about creating a custom index component?

**Reply by Paul-Bob** [reply]
Hi @manuelmeurer,

https://docs.avohq.io/3.0/field-options.html#components

The field components docs show how to eject and override existing components.

You can follow that guide to eject a component using a scope, modify it, and apply it to all `textarea` fields via the `components` option. But in this case, there's a better approach. I'll open one issue to document this use case and another issue with a proposal to create a dedicated `TextareaField::IndexComponent` and discuss it for Avo4.

Since the `textarea` field doesn't have an index component, you can create one.

Instead of starting from scratch, it's easier to use the `text` field’s index component as a base. Here's how you can do that in this guide:


1. Run:
   ```sh
   rails g avo:eject --field-components text --view index
   ```

2. Rename the generated directory from `text_field` to `textarea_field`.

3. Update the class reference from:
   ```ruby
   Avo::Fields::TextField::IndexComponent
   ```
   to:
   ```ruby
   Avo::Fields::TextareaField::IndexComponent
   ```

4. Replace the ERB content with something like:
   ```erb
   <%= index_field_wrapper(**field_wrapper_args) do %>
     <%= @field.value.truncate(60) %>
   <% end %>
   ```

5. Wait for [this PR](https://github.com/avo-hq/avo/pull/3830) to be merged and released, currently, `textarea` fields are forced to be hidden on index.  
   If you need it right away, you can monkey patch the PR changes.

6. Once the PR is merged, go to the relevant `textarea` fields and add:
   ```ruby
   show_on: :index
   ```

Step 6 is needed because `textarea` fields are currently forced to be hidden on the index view.

[This PR](https://github.com/avo-hq/avo/pull/3830) will allow overriding that behavior.

If you want to force all `textarea` fields to use `show_on: :index`, you can extend the base resource and override the `field` method to apply that option globally.

Docs for extending the base resource: https://docs.avohq.io/3.0/resources.html#extending-avo-baseresource

Example:

```ruby
# app/avo/base_resource.rb
module Avo
  class BaseResource < Avo::Resources::Base
    # Example custom method: make all number fields cast their values to float
    def field(id, **args, &block)
      if args[:as] == :textarea
        args[:show_on] = :index
      end

      super(id, **args, &block)
    end
  end
end
```

Really appreciate your patience, just a heads-up that issues coming through [support subscriptions](https://avohq.io/support) are always prioritized

**Reply by Paul-Bob** [reply]
Linking the related issues:

- https://github.com/avo-hq/avo/discussions/3764#discussioncomment-13038906
- https://github.com/avo-hq/docs.avohq.io/issues/380

**Comment by xeron** [comment]
Images in `files` field get `aspect-video` class applied which stretches vertical images. Feels like even just completely removing `aspect-video` already looks better:

<img width="1259" height="223" alt="1" src="https://github.com/user-attachments/assets/8aeb1f75-bd26-411e-aa3f-f9cc1639d69c" />
<img width="1262" height="382" alt="2" src="https://github.com/user-attachments/assets/2c0794f8-3670-48f4-92ee-10adc703f5ab" />

**Reply by Paul-Bob** [reply]
Thanks for surfacing this issue!
